src/components/lab/NumberInput/utils.ts

```diff
@@ -0,0 +1,262 @@
+
+    return {valid: isValidValue, value: parsedValue};
+}
+
+function roundIfNecessary(value: number, allowDecimal: boolean) {
+    return allowDecimal ? value : Math.floor(value);
+}
+
+interface VariablesProps {
+    min: number | undefined;
+    max: number | undefined;
+    step: number;
+    shiftMultiplier: number;
+    value: number | undefined;
+    defaultValue: number | undefined;
+    allowDecimal: boolean;
+}
+export function getInternalVariables(props: VariablesProps): {
+    min: number | undefined;
+    max: number | undefined;
+    step: number;
+    shiftMultiplier: number;
+    value: number;
+    defaultValue: number;
+} {
+    warnAboutInvalidProps(props);
+
+    const {
+        min: externalMin,
+        max: externalMax,
+        step: externalStep,
+        shiftMultiplier: externalShiftMultiplier,
+        value: externalValue = 0,
+        allowDecimal,
+        defaultValue: externalDefaultValue = 0,
+    } = props;
+
+    const {min: rangedMin, max: rangedMax} =
+        externalMin && externalMax && externalMin > externalMax
+            ? {
+                  min: externalMax,
+                  max: externalMin,
+              }
+            : {min: externalMin, max: externalMax};
+
+    const min = rangedMin && rangedMin >= Number.MIN_SAFE_INTEGER ? rangedMin : undefined;
+    const max = rangedMax && rangedMax <= Number.MAX_SAFE_INTEGER ? rangedMax : undefined;
+
+    const step = roundIfNecessary(Math.abs(externalStep), allowDecimal) || 1;
+    const shiftMultiplier = roundIfNecessary(externalShiftMultiplier, allowDecimal) || 10;
+    const value = roundIfNecessary(externalValue, allowDecimal);
+    const defaultValue = roundIfNecessary(externalDefaultValue, allowDecimal);
+
+    return {min, max, step, shiftMultiplier, value, defaultValue};
+}
+
+export function clampToNearestStepValue({
+    value,
+    step,
+    min: originalMin,
+    max = Number.MAX_SAFE_INTEGER,
+}: {
+    value: number;
+    step: number;
+    min: number | undefined;
+    max: number | undefined;
+}) {
+    const base = originalMin || 0;
+    const min = originalMin ?? Number.MIN_SAFE_INTEGER;
+    const clampedValue = Math.max(min, Math.min(max, value));
+    const stepDeviation = (clampedValue - base) % step;
+    const amountOfStepsDiff = Math.floor((clampedValue - base) / step);
+    if (stepDeviation !== 0) {
+        const smallerPossibleValue = base + amountOfStepsDiff * step;
+        const greaterPossibleValue = base + (amountOfStepsDiff + 1) * step;
+
+        if (
+            (greaterPossibleValue > max ||
+                greaterPossibleValue - clampedValue > clampedValue - smallerPossibleValue) &&
+            smallerPossibleValue >= min
+        ) {
+            return smallerPossibleValue;
+        }
+        if (greaterPossibleValue <= max) {
+            return greaterPossibleValue;
+        }
+    }
+    return clampedValue;
+}
+
+export function warnAboutInvalidProps({
```

Comment:
This will span a lot. What's the benefit of this? Component won't allow invalid value anyway.

Reply:
It will spam only with invalid props passed.
I think it would be better to notify programmer about invalid component usage6 but not just fix props silently